### DISCUSSION TITLE
Advice: Migration 1.3->1.4 Issue -- "Can't Adapt Type Row"; wherever query(X.y).one() uses "y" (postgres specific?)

### DISCUSSION THREAD

**Discussion by maquino1985** [sqlalchemy/sqlalchemy#9214] [Usage Questions] [answered]
Hi, I hope the title is not too confusing (I can rephrase if you have a better suggestion!)

I am upgrading a large codebase from 1.3 to 1.4 (prepping for 2.0) and everywhere we queried for, e.g. an ID the return object is now a Row and causes postgres/SQLA to throw an error: sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) can't adapt type 'Row'

1.3 compatible code:
```python
cv_uop_type_id = (
            session.query(CvUnitOperationType.id)
            .filter(CvUnitOperationType.entry_key == operation_type_enum.value)
            .one()
        )
        uop = UnitOperation(
            cv_unit_operation_type_id=cv_uop_type_id,
            user_id=user_id,
            additional_information=additional_information,
        )
```

in 1.4 this needs to be changed to 
```python
        uop = UnitOperation(
            cv_unit_operation_type_id=cv_uop_type_id.id, # the change is cv_uop_type_id->cv_uop_type_id.id
            user_id=user_id,
            additional_information=additional_information,
        )
```

This is is used _a lot_ in our codebase, and I'm worried about trying to manually find every instance. Is there a better/recommended approach? I tried using the register_adapter approach suggested in https://gaganpreet.in/posts/sqlalchemy-cant-adapt-type/ but that did not work.

**Comment by zzzeek** [comment, marked as answer]
I'm not sure what your code does with 1.3, the return type of `one()` in this case is not an integer in either SQLAlchemy 1.3 **or** 1.4, in both cases it's a SQLAlchemy object that acts like a named tuple.

Here's a demo:

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import Session

Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)

e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add(A())
s.commit()

this_is_a_row = s.query(A.id).one()
print(type(this_is_a_row))
print(this_is_a_row.id)
```

in 1.3 we have:

```
<class 'sqlalchemy.util._collections.result'>
1
```

in 1.4 we have:

```
<class 'sqlalchemy.engine.row.Row'>
1
```

the "result" and "Row" are 99% compatible.  so you might have code in your app somewhere that is doing `isinstance(obj, sqlalchemy.util._collections.result)` or something like that, otherwise I can't see how the 1.3 code has some better result without calling upon ``.id``.

**Reply by maquino1985** [reply]
I think it's simply that the Row can't be used where the Result could be by the postgres driver? 

Here's another example where I am fixing 1.3 code to work with 1.4
<img width="737" alt="image" src="https://user-images.githubusercontent.com/13229385/216142505-7d29c916-4e44-42a7-bb72-3a92f1a03bfe.png">

1.4 
```python
 organization_ids = g.session.query(CvOrganization.id).all()
```

[(UUID('4c786add-a7b9-44c1-8cee-ad87883db340'),), (UUID('5b36a105-5410-4c55-8c08-31f77a75c11b'),), (UUID('e71fd020-5337-4b36-846e-7b11202d7949'),), (UUID('fb3df8ce-dd7c-46a3-afa5-fd71ce4fcc23'),), (UUID('8f81048b-8cf8-46f8-b1a9-e3b537631b29'),)]

<img width="586" alt="image" src="https://user-images.githubusercontent.com/13229385/216143124-4a986100-d8f2-4ce1-b3d2-7085f8390d15.png">


1.3
<img width="575" alt="image" src="https://user-images.githubusercontent.com/13229385/216143903-a3a57138-7060-4751-874f-559f1ad61903.png">

<img width="577" alt="image" src="https://user-images.githubusercontent.com/13229385/216144155-15c36102-f9ee-4cf0-9d3a-6ca827048b24.png">

print(organization_ids) looks the same as in 1.4, however: [(UUID('4c786add-a7b9-44c1-8cee-ad87883db340'),), (UUID('5b36a105-5410-4c55-8c08-31f77a75c11b'),), (UUID('e71fd020-5337-4b36-846e-7b11202d7949'),), (UUID('fb3df8ce-dd7c-46a3-afa5-fd71ce4fcc23'),), (UUID('8f81048b-8cf8-46f8-b1a9-e3b537631b29'),)]

I don't know specifically what the driver does, but the list of tuples requires that you unpack the value e.g. row.id or row[0] whereas the result object just let you use it and was auto unpacked.


If you'd like a working example I can create one definitely, but I think just writing a query in 1.3 that uses this syntax and then changing to 1.4 will cause the same issue with the postgres driver.

**Reply by CaselIT** [reply]
I think what you are seeing are just different str representation.

the last image clearly show that result is a tuple like of len 1

**Reply by maquino1985** [reply]
> I think what you are seeing are just different str representation.
> 
> the last image clearly show that result is a tuple like of len 1

Yes, that is true. That's not the point I was trying to make, sorry if it was unclear. it's simply that in 1.3 the "results" could be used as if they were scalars/lists of scalars whereas in 1.4 the "rows" need to be treated like Lists of NamedTuples so everywhere that we have anything that used a result in that way needs to be refactored or psycopg yells that it cannot adapt Row.

If you do something like

```python
from sqlalchemy import Column, ForeignKey
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import Session

Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    b_fk = Column(Integer, ForeignKey("b.id"))

class B(Base):
    __tablename__ = 'b'

    id = Column(Integer, primary_key=True)

e = create_engine(
    "postgresql://postgres:postgres@localhost:5432/dbname", echo=True
)
Base.metadata.create_all(e)

s = Session(e)

s.add(B())
s.commit()

b_id = s.query(B.id).one()
a = A(b_fk=b_id)
s.add(a)
s.commit() # sqlalchemy.exc.InterfaceError: (sqlite3.InterfaceError) Error binding parameter 0 - probably unsupported type. (if you use sqlite) or sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) can't adapt type 'Row' (if you use postgres)
```

In 1.3, this will commit:
s.commit()
2023-02-01 15:03:54,654 INFO sqlalchemy.engine.base.Engine INSERT INTO a (b_fk) VALUES (%(b_fk)s) RETURNING a.id
2023-02-01 15:03:54,654 INFO sqlalchemy.engine.base.Engine {'b_fk': (1,)}
2023-02-01 15:03:54,656 INFO sqlalchemy.engine.base.Engine COMMIT


_The error occurs in sqlite but NOT postgres in 1.3_ (ugh!)

**Reply by zzzeek** [reply]
So this is the psycopg2 driver being given the value `(1, )`, a one-element tuple, as a parameter that is ultimately bound to a SQL string given the way that psycopg2 binds parameters.     psycopg2 seems to just accept tuples and silently flattens them into their lone element, unless you pass more than one element:

```py
import psycopg2

conn = psycopg2.connect(
    user="scott", password="tiger", database="test", host="localhost"
)

cur = conn.cursor()

cur.execute("delete from a")
cur.execute(
    "insert into a (id, data) values (%(id)s, %(data)s)",
    {"id": (1,), "data": ("some data",)},
)
cur.execute(
    "insert into a (id, data) values (%(id)s, %(data)s)",
    {"id": (2,), "data": ("some data", "longer tuple")},
)
cur.execute("select * from a")
print(cur.fetchall())

```

returns:

```
[(1, None, 'some data'), (2, None, '("some data","longer tuple")')]
```

unfortunately psycopg2 is simply wrong here and your program has been relying on a driver bug.   I think fixing this issue in your codebase has to happen regardless.

**Reply by maquino1985** [reply]
Thanks. I really appreciate your time and the assistance. While that's certainly not the outcome I was hoping for, it does make sense.
